src/tron/tronEngine.ts

```diff
@@ -1043,6 +1047,95 @@ export class TronEngine extends CurrencyEngine<TronTools, SafeTronWalletInfo> {
     return edgeTransaction
   }
 
+  async makeFreezeV2Transaction(
+    action: TronFreezeV2Action
+  ): Promise<EdgeTransaction> {
+    const {
+      params: { nativeAmount, resource }
+    } = action
+
+    const contractJson = {
+      parameter: {
+        value: {
+          owner_address: base58ToHexAddress(this.walletLocalData.publicKey),
+          frozen_balance: parseInt(nativeAmount),
+          resource: resource === 'ENERGY_V2' ? 'ENERGY' : 'BANDWIDTH'
+        }
+      },
+      type: 'FreezeBalanceV2Contract'
+    }
+
+    const txOtherParams: TxBuilderParams = { contractJson }
+    const { transactionHex } = await this.txBuilder(txOtherParams)
+    const networkFee = await this.calcTxFee({ unsignedTxHex: transactionHex })
+
+    const edgeTransaction: EdgeTransaction = {
+      txid: '',
+      date: 0,
+      currencyCode: this.currencyInfo.currencyCode,
+      blockHeight: 0,
+      nativeAmount: mul(nativeAmount, '-1'),
+      isSend: true,
+      networkFee,
+      ourReceiveAddresses: [],
+      signedTx: '',
+      otherParams: txOtherParams,
+      walletId: this.walletId,
+      metadata: {
+        notes: resource
+      }
+    }
+
+    return edgeTransaction
+  }
+
+  async makeUnfreezeV2Transaction(
+    action: TronUnfreezeV2Action
+  ): Promise<EdgeTransaction> {
+    const {
+      params: { nativeAmount, resource }
+    } = action
+
+    const stakedAmount = this.stakingStatus.stakedAmounts.find(
+      amount => amount.otherParams?.type === resource
+    )
+    if (stakedAmount == null) throw new Error('Nothing to unfreeze')
+
+    const contractJson = {
+      parameter: {
+        value: {
+          owner_address: base58ToHexAddress(this.walletLocalData.publicKey),
+          unfreeze_balance: parseInt(nativeAmount),
```

Comment:
How do `nativeAmount` and `stakedAmount` work together? I feel like I am missing something here, because one is going to the Tron SDK, and the other is being returned to Edge.